### PR TITLE
Feature: Transition to celery for background tasks

### LINKED ISSUE #1321: [BUG] Bare metal install of django-q fails
### Description

I'm receiving an error when installing requirements for v1.8.0 in a bare metal installation specific to the line for django-q.

 -e git+https://github.com/paperless-ngx/django-q.git@bf20d57f859a7d872d5979cd8879fac9c9df981c#egg=django-q

```
Obtaining django-q from git+https://github.com/paperless-ngx/django-q.git@bf20d57f859a7d872d5979cd8879fac9c9df981c#egg=django-q
ERROR: Exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/_internal/vcs/git.py", line 299, in get_remote_url
    found_remote = remotes[0]
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/_internal/cli/base_command.py", line 186, in _main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/_internal/commands/install.py", line 357, in run
    resolver.resolve(requirement_set)
  File "/usr/lib/python3/dist-packages/pip/_internal/legacy_resolve.py", line 177, in resolve
    discovered_reqs.extend(self._resolve_one(requirement_set, req))
  File "/usr/lib/python3/dist-packages/pip/_internal/legacy_resolve.py", line 333, in _resolve_one
    abstract_dist = self._get_abstract_dist_for(req_to_install)
  File "/usr/lib/python3/dist-packages/pip/_internal/legacy_resolve.py", line 265, in _get_abstract_dist_for
    return self.preparer.prepare_editable_requirement(req)
  File "/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py", line 552, in prepare_editable_requirement
    req.update_editable(not self._download_should_save)
  File "/usr/lib/python3/dist-packages/pip/_internal/req/req_install.py", line 654, in update_editable
    vcs_backend.obtain(self.source_dir, url=hidden_url)
  File "/usr/lib/python3/dist-packages/pip/_internal/vcs/versioncontrol.py", line 532, in obtain
    existing_url = self.get_remote_url(dest)
  File "/usr/lib/python3/dist-packages/pip/_internal/vcs/git.py", line 301, in get_remote_url
    raise RemoteNotFoundError
pip._internal.vcs.versioncontrol.RemoteNotFoundError
```

### Steps to reproduce

Pull v1.8.0
sudo -Hu paperless pip3 install -r requirements.txt

### Webserver logs

_No response_

### Paperless-ngx version

1.8.0

### Host OS

Ubunt 20.04

### Installation method

Bare metal

### Browser

Chrome

### Configuration changes

_No response_

### Other

_No response_

### LINKED ISSUE #1355: [BUG] InterfaceError while consuming document
### Description

Since Version 1.80 all kind of documents (docx, PDF, etc) wich i put in my consume folder executing the consuming process stops with the following error:

### Steps to reproduce

Insert a document into the consume folder.

### Webserver logs

```bash
[2022-08-03 21:10:54,601] [INFO] [paperless.management.consumer] Adding /usr/src/paperless/consume/TEST.docx to the task queue.
[2022-08-03 21:10:54,604] [ERROR] [paperless.management.consumer] Error while consuming document
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/base.py", line 262, in _cursor
    return self._prepare_cursor(self.create_cursor(name))
  File "/usr/local/lib/python3.9/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/postgresql/base.py", line 256, in create_cursor
    cursor = self.connection.cursor()
psycopg2.InterfaceError: connection already closed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/paperless/src/documents/management/commands/document_consumer.py", line 94, in _consume
    async_task(
  File "/usr/src/paperless/src/docker/src/django-q/django_q/tasks.py", line 67, in async_task
    pre_enqueue.send(sender="django_q", task=task)
  File "/usr/local/lib/python3.9/site-packages/django/dispatch/dispatcher.py", line 176, in send
    return [
  File "/usr/local/lib/python3.9/site-packages/django/dispatch/dispatcher.py", line 177, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/usr/src/paperless/src/documents/signals/handlers.py", line 511, in init_paperless_task
    paperless_task, created = PaperlessTask.objects.get_or_create(
  File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 657, in get_or_create
    return self.get(**kwargs), False
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 492, in get
    num = len(clone)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 302, in __len__
    self._fetch_all()
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 1507, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 57, in __iter__
    results = compiler.execute_sql(
  File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1359, in execute_sql
    cursor = self.connection.cursor()
  File "/usr/local/lib/python3.9/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/base.py", line 284, in cursor
    return self._cursor()
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/base.py", line 262, in _cursor
    return self._prepare_cursor(self.create_cursor(name))
  File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/base.py", line 262, in _cursor
    return self._prepare_cursor(self.create_cursor(name))
  File "/usr/local/lib/python3.9/site-packages/django/utils/asyncio.py", line 26, in inner
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/postgresql/base.py", line 256, in create_cursor
    cursor = self.connection.cursor()
django.db.utils.InterfaceError: connection already closed
```

### Paperless-ngx version

1.80

### Host OS

Synology NAS

### Installation method

Docker - official image

### Browser

_No response_

### Configuration changes

no changes since 1.7.0

### Other

_No response_

### PR DESCRIPTION
**Pull request by stumpylog**
## Proposed change

With `django-q` remaining unmaintained and myself being the main keeper of our own copy of it, it is probably best to transition away from it and to something supported, documented, etc.

This PR transitions the backend to use [celery](https://docs.celeryq.dev/en/stable/index.html) and celery beat for background tasks and scheduled tasks.  It is maintained, has extensive documentation, at least some Django integrations and generally a lot of features.  

I have successfully migrated a 1.8.0 Docker instance to this image, so I don't believe this is a breaking change.  For bare metal installs, it will require some more work, but I believe that is already the case.  All the tasks, including scheduled tasks are working.  The frontend task view is not yet working

### TODOs
- [x] Get frontend File Tasks view working
- [x] Fix frontend testing
- [x] [Update documentation](https://github.com/search?q=qcluster+repo%3Apaperless-ngx%2Fpaperless-ngx+filename%3A*.rst&type=Code&ref=advsearch&l=&l=)
- [x] Extensive testing

### Testing
- Stress Test
  - Consumed ~1000 files
  - postgres as the db
  - 16 workers.
  - All files were from an export
  - All files consumed, frontend tasks updated as expected
- In-place Upgrade
  - 1.7 -> this branch, no issues
  - 1.8 -> this branch, no issues
  - 1.9 -> this branch, no issues

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by shamoon** [comment]
Do you think its possible to migrate the frontend task queue without that unmerged PR you linked? I dont pretend to know the library off-hand but I see celery [makes good use of signals](https://docs.celeryq.dev/en/stable/userguide/signals.html) so could we transition the q signals to those?

**Comment by stumpylog** [comment]
It could be, but I was considering the number of odd database problems we had around using signals and not really wanting to have those happening again.  If that PR gets merged, then the code for creating a model in the right times is in django-celery-results, while we would only need to extend it with the `acknowledged` field.  Basically, less for us to manage and do.

It's possible celery is better behaved with database connections and signals, so I'll look into it more.  

The view to tasks is the last large part still TODO.  Once I have something in the backend, I'll need your expert skills to fix my damage to the frontend.

**Comment by shamoon** [comment]
Gotcha. Of course I would be happy to help, feel free to break the frontend. I will take a look at celery's signal stuff when I have some time, it took me a while to figure it out with q so I have a feeling you (or someone else) will get there first. Either way obviously I defer to you on the best way to go about it.

**Comment by stumpylog** [comment]
I hooked into the TaskResult post_save signal instead.  It's the closest thing to abstract away a lot of the when to save/update our model.  The JSON looks pretty close now, but there's a few differences, as the fields were directly one to one.

- `created` is `date_created`
- No real equivalent to `started`, closest is `date_done`
- Status is now one of `{'STARTED', 'RETRY', 'SUCCESS', 'REVOKED', 'RECEIVED', 'PENDING', 'FAILURE'}`
  - if it's helpful, the backend could probably condense some of those

So maybe when you've got some time, you could work on the frontend portion needed for the differences

**Comment by shamoon** [comment]
That was fast! Ok I'll take a look, one thing as I was just firing things up to poke around is when I try to migrate my test install I think the migrations dont work if `attempted_task` isnt null, e.g. Im getting:
```
django.db.utils.IntegrityError: The row in table 'documents_paperlesstask' with primary key '11' has an invalid foreign key: documents_paperlesstask.attempted_task_id contains a value '5f8e8ed5e36242639f52405e5a034803' that does not have a corresponding value in django_q_task.id
```
presumably because we dropped the q tables already. Do we want to null-out those values first?

**Comment by stumpylog** [comment]
Oh, I see.  I re-ordered the migrations, but I forgot I did that order for a reason.  One sec

**Comment by shamoon** [comment]
Ok I think I got the frontend part squared away! I also took the opportunity to add a little feature where we link directly to a doc if its a successful consumption, e.g.:
related

Of course feel free to modify what I did, I did have to do some funny things like wrangle the `task_kwargs` content into a proper JSON object string, etc.

**Comment by stumpylog** [comment]
Yeah, that's some weird storage of task_args and task_kwargs.  It's like a double string of a Python container.  I found a few open issues in celery and celery-results about it, but nothing merged or fixed.  I changed it up a little bit and added testing to cover the cases I could think of.

The document open is a nice touch.

File: src/documents/serialisers.py [hunk 1]
```diff
@@ -632,24 +652,108 @@ def get_type(self, obj):
...
+                if not len(result):
+                    # There are always some arguments to the consume
+                    task_args: Tuple = literal_eval(
+                        literal_eval(obj.attempted_task.task_args),
+                    )
+                    filepath = Path(task_args[0])
+                    result = filepath.name
+            except Exception as e:  # pragma: no cover
+                # Extra security if something is malformed
+                logger.warn(f"Error getting task result: {e}", exc_info=True)
```
Comment (shamoon) [inline src/documents/serialisers.py:736]:
```suggestion
                logger.warn(f"Error getting file name from task: {e}", exc_info=True)
```

=)

**Comment by stumpylog** [comment]
Alright, I've tweaked the documentation and added the necessary systemd fie.  I want to test out more just to be sure about upgrading in place (migrations in particular), but I think this is fully ready now.

**Comment by stumpylog** [comment]
This will definitely fix #1321 and I hope fix #1355

**Review by shamoon — APPROVED** [review]
Not sure if the current test failures are real (they look weird, Im just re-running) but I think this is great and I vote we forge ahead as discussed in the team thread as well

**Comment by shamoon** [comment]
Ah
```
Error: Unable to locate the previous sha: 71b45715244dd036948e7cf38307e7f29c7142db
[252](https://github.com/paperless-ngx/paperless-ngx/actions/runs/3177304226/jobs/5179989679#step:9:256)
  Error: You seem to be missing 'fetch-depth: 0' or 'fetch-depth: 2'. See https://github.com/tj-actions/changed-files#usage
[253](https://github.com/paperless-ngx/paperless-ngx/actions/runs/3177304226/jobs/5179989679#step:9:257)
  Error: Process completed with exit code 1.
```
maybe from #1702?

**Comment by stumpylog** [comment]
They're not really. The tests pass, it's a problem with the changed files action.  Don't think it likes me rebasing

**Comment by shamoon** [comment]
Passing now 😎

**Comment by shamoon** [comment]
Alright lets do it, clearly next release will need some good beta testing =)

**Pull request merged**